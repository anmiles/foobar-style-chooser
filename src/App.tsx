import './styles/App.css';

import { useState } from 'react';

import sections from '../config/sections.json'; // eslint-disable-line n/no-unpublished-import

import Result from './Result';
import Section from './Section';

export default function App() {
	const [ selectedStyles, setSelectedStyles ] = useState<string[]>([]);

	function handleStyleClick(style: string) {
		const nextSelectedStyles = selectedStyles.includes(style)
			? selectedStyles.filter((s) => s !== style)
			: [ ...selectedStyles, style ];

		setSelectedStyles(nextSelectedStyles);
	}

	return (
		<div
			className="app"
			data-testid="app"
		>
			{ sections.map((sectionData) => (
				'isHidden' in sectionData && sectionData.isHidden
					? null
					: (
							<Section
								key={ sectionData.title }
								selectedStyles={ selectedStyles }
								handleStyleClick={ handleStyleClick }
								{ ...sectionData }
							/>
						)
			),
			) }

			<Result
				selectedStyles={ selectedStyles }
				setSelectedStyles={ setSelectedStyles }
			/>
		</div>
	);
}
