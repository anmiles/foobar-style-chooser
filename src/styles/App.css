.app {
	width: 100%;
	min-height: 100%;
	flex-direction: column;
	display: flex;
	gap: 16px;
	padding: 16px;
}

.app > * {
	flex-direction: row;
	flex-wrap: wrap;
	border: 2px solid currentColor;
	border-radius: 1em;
	position: relative;
	display: grid;
	grid-column-gap: 16px;
	grid-row-gap: 16px;
	padding: 16px;
	grid-template-columns: repeat(12, minmax(0, 1fr));
}
