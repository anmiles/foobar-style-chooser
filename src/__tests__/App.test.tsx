import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import App from '../App';

jest.mock('../../config/sections.json', () => require('./sections.test.json'));

beforeEach(() => {
	render(<App />);
});

describe('src/App', () => {
	describe('section', () => {
		it('should be correctly rendered', () => {
			const sectionElements = screen.getAllByRole('section');

			expect(sectionElements).toMatchSnapshot();
		});

		it('should toggle style selection on click', async () => {
			const style = screen.getByTestId('style-piano');

			expect(style).toMatchSnapshot();

			await userEvent.click(style);

			expect(style).toMatchSnapshot();

			await userEvent.click(style);

			expect(style).toMatchSnapshot();
		});
	});

	describe('result', () => {
		it('should be rendered', () => {
			const resultElement = screen.getByTestId('result');

			expect(resultElement).toMatchSnapshot();
		});

		describe('value', () => {
			it('should contain no styles', () => {
				const valueElement = screen.getByTestId('value');

				expect(valueElement.textContent).toEqual(':');
			});

			it('should contain selected styles ordered alphabetically', async () => {
				await userEvent.click(screen.getByTestId('style-classical'));
				await userEvent.click(screen.getByTestId('style-active'));
				await userEvent.click(screen.getByTestId('style-relax'));
				await userEvent.click(screen.getByTestId('style-active'));
				await userEvent.click(screen.getByTestId('style-piano'));

				const valueElement = screen.getByTestId('value');

				expect(valueElement.textContent).toEqual(':Classical:Piano:Relax:');
			});
		});

		describe('buttons', () => {
			const readText  = jest.fn();
			const writeText = jest.fn();

			beforeEach(() => {
				Object.defineProperty(navigator, 'clipboard', {
					value: {
						readText,
						writeText,
					},
					writable: true,
				});
			});

			describe('copy', () => {
				it('should copy sorted result in clipboard', async () => {
					await userEvent.click(screen.getByTestId('style-classical'));
					await userEvent.click(screen.getByTestId('style-active'));
					await userEvent.click(screen.getByTestId('style-relax'));
					await userEvent.click(screen.getByTestId('style-active'));
					await userEvent.click(screen.getByTestId('style-piano'));

					await userEvent.click(screen.getByTestId('copy'));

					expect(writeText).toHaveBeenCalledWith(':Classical:Piano:Relax:');
				});
			});

			describe('paste', () => {
				beforeEach(() => {
					readText.mockResolvedValue(':Unknown:Piano::Relax:Classical:');
				});

				it('should paste sorted result from clipboard', async () => {
					await userEvent.click(screen.getByTestId('paste'));

					await waitFor(() => {
						const valueElement = screen.getByTestId('value');
						expect(valueElement.textContent).toEqual(':Classical:Piano:Relax:Unknown:');
					});
				});

				it('should apply style selections', async () => {
					await userEvent.click(screen.getByTestId('paste'));

					const sectionElements = screen.getAllByRole('section');

					expect(sectionElements).toMatchSnapshot();
				});
			});
		});
	});
});

