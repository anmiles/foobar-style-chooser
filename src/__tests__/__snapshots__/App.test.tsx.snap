// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/App result buttons paste should apply style selections 1`] = `
[
  <div
    class="section genres"
    data-testid="section-genres"
    role="section"
  >
    <a
      class="style button isSelected"
      data-testid="style-classical"
      role="style"
    >
      Classical
    </a>
    <a
      class="style button"
      data-testid="style-avantgarde"
      role="style"
    >
      Avantgarde
    </a>
  </div>,
  <div
    class="section instruments"
    data-testid="section-instruments"
    role="section"
  >
    <a
      class="style button isSelected"
      data-testid="style-piano"
      role="style"
    >
      Piano
    </a>
    <a
      class="style button"
      data-testid="style-electronic"
      role="style"
    >
      Electronic
    </a>
  </div>,
  <div
    class="section themes"
    data-testid="section-themes"
    role="section"
  >
    <a
      class="style button"
      data-testid="style-active"
      role="style"
    >
      Active
    </a>
    <a
      class="style button isSelected"
      data-testid="style-relax"
      role="style"
    >
      Relax
    </a>
  </div>,
]
`;

exports[`src/App result should be rendered 1`] = `
<div
  class="result"
  data-testid="result"
>
  <a
    data-testid="value"
  >
    :
  </a>
  <input
    class="button"
    data-testid="copy"
    type="button"
    value="Copy"
  />
  <input
    class="button"
    data-testid="paste"
    type="button"
    value="Paste"
  />
</div>
`;

exports[`src/App section should be correctly rendered 1`] = `
[
  <div
    class="section genres"
    data-testid="section-genres"
    role="section"
  >
    <a
      class="style button"
      data-testid="style-classical"
      role="style"
    >
      Classical
    </a>
    <a
      class="style button"
      data-testid="style-avantgarde"
      role="style"
    >
      Avantgarde
    </a>
  </div>,
  <div
    class="section instruments"
    data-testid="section-instruments"
    role="section"
  >
    <a
      class="style button"
      data-testid="style-piano"
      role="style"
    >
      Piano
    </a>
    <a
      class="style button"
      data-testid="style-electronic"
      role="style"
    >
      Electronic
    </a>
  </div>,
  <div
    class="section themes"
    data-testid="section-themes"
    role="section"
  >
    <a
      class="style button"
      data-testid="style-active"
      role="style"
    >
      Active
    </a>
    <a
      class="style button"
      data-testid="style-relax"
      role="style"
    >
      Relax
    </a>
  </div>,
]
`;

exports[`src/App section should toggle style selection on click 1`] = `
<a
  class="style button"
  data-testid="style-piano"
  role="style"
>
  Piano
</a>
`;

exports[`src/App section should toggle style selection on click 2`] = `
<a
  class="style button isSelected"
  data-testid="style-piano"
  role="style"
>
  Piano
</a>
`;

exports[`src/App section should toggle style selection on click 3`] = `
<a
  class="style button"
  data-testid="style-piano"
  role="style"
>
  Piano
</a>
`;
